types/google-apps-script/google-apps-script-tests.ts

```diff
@@ -31,178 +31,332 @@ CalendarApp.GuestStatus.NO;
-  // Create file.
-  const file = DriveApp.createFile('New Text File', 'Hello, world!');
-  // Get description. Expect null.
-  Logger.log(file.getDescription());
-  // Set description.
-  file.setDescription('desc');
-  // Get description. Expect 'DESC'.
-  Logger.log(file.getDescription().toUpperCase());
+    // Create file.
+    const file = DriveApp.createFile("New Text File", "Hello, world!");
+    // Get description. Expect null.
+    Logger.log(file.getDescription());
+    // Set description.
+    file.setDescription("desc");
+    // Get description. Expect 'DESC'.
+    Logger.log(file.getDescription().toUpperCase());
 };
 
 function timeDriven(e: GoogleAppsScript.Events.TimeDriven) {
-  if (typeof e.hour === 'number') {
-      console.log('Time driven event detected');
-  }
+    if (typeof e.hour === "number") {
+        console.log("Time driven event detected");
+    }
 }
 
 CardService.newDecoratedText(); // $ExpectType DecoratedText
 CardService.newDecoratedText().setAuthorizationAction(CardService.newAuthorizationAction()); // $ExpectType DecoratedText
-CardService.newDecoratedText().setBottomLabel(''); // $ExpectType DecoratedText
+CardService.newDecoratedText().setBottomLabel(""); // $ExpectType DecoratedText
 CardService.newDecoratedText().setButton(CardService.newTextButton()); // $ExpectType DecoratedText
 CardService.newDecoratedText().setComposeAction(CardService.newAction(), CardService.ComposedEmailType.REPLY_AS_DRAFT); // $ExpectType DecoratedText
 CardService.newDecoratedText().setIcon(CardService.Icon.AIRPLANE); // $ExpectType DecoratedText
-CardService.newDecoratedText().setIconAltText(''); // $ExpectType DecoratedText
-CardService.newDecoratedText().setIconUrl(''); // $ExpectType DecoratedText
+CardService.newDecoratedText().setIconAltText(""); // $ExpectType DecoratedText
+CardService.newDecoratedText().setIconUrl(""); // $ExpectType DecoratedText
 CardService.newDecoratedText().setOnClickAction(CardService.newAction()); // $ExpectType DecoratedText
 CardService.newDecoratedText().setOnClickOpenLinkAction(CardService.newAction()); // $ExpectType DecoratedText
 CardService.newDecoratedText().setOpenLink(CardService.newOpenLink()); // $ExpectType DecoratedText
 CardService.newDecoratedText().setSwitchControl(CardService.newSwitch()); // $ExpectType DecoratedText
-CardService.newDecoratedText().setText(''); // $ExpectType DecoratedText
-CardService.newDecoratedText().setTopLabel(''); // $ExpectType DecoratedText
+CardService.newDecoratedText().setText(""); // $ExpectType DecoratedText
+CardService.newDecoratedText().setTopLabel(""); // $ExpectType DecoratedText
 CardService.newDecoratedText().setWrapText(true); // $ExpectType DecoratedText
 
-DriveApp.createShortcut('').getTargetId();
-DriveApp.createFile('', '').moveTo(DriveApp.getFolderById(''));
+DriveApp.createShortcut("").getTargetId();
+DriveApp.createFile("", "").moveTo(DriveApp.getFolderById(""));
+
+// Addon event objects tests:
```

Comment:
@mtgto thanks! Frankly, I forgot to add a test specifically for it - will add during the weekend

Reply:
> I found your test code does not contain `CommonEventObject`, but your test is good enough 👍

@mtgto - done, the `CommonEventObject` should now be covered by a test as well. Also noticed that I did not make the `formInputs[string][""]` optional because only one of the object structures is present at a time - addressed (I wanted to add a generic `OneOf` helper, but wasn't sure if that's ok given that Apps Script declaration files do not use helpers - please, let me know if you think it is a good or bad idea to add one).

Reply:
> done, the `CommonEventObject` should now be covered by a test as well. 

Agree. Your test is very well 👍 👍  Nice work!

> Also noticed that I did not make the `formInputs[string][""]` optional because only one of the object structures is present at a time

It seems that you don't need to define using such a `OneOf` helper. It's work for user which uses this PR's definition, IMO.
For example, it is inconvenient to use these type:

```typescript
interface CommonEventObject {
    formInputs: {
        [ID: string]: {
            "":
                | { stringInputs: StringInputObject }
                | { dateInput: DateInputObject }
                | { timeInput: TimeInputObject }
                | { dateTimeInput: DateTimeInputObject };
        };
    };
}

const f = (foo: GoogleAppsScript.AddonEventObjects.CommonEventObject) => {
    const bar = foo.formInputs.employeeName[""];
    if (bar.stringInputs) {
        // ...
    }
};
```

function f causes compile error:
```
Property 'stringInputs' does not exist on type '{ stringInputs: StringInputObject; } | { dateInput: DateInputObject; } | { timeInput: TimeInputObject; } | { dateTimeInput: DateTimeInputObject; }'.
  Property 'stringInputs' does not exist on type '{ dateInput: DateInputObject; }'.ts(2339)
```

Also I can't find any document of `commonEventObject.formInputs.xxxx[""]` has exact one entry only.
https://developers.google.com/workspace/add-ons/concepts/event-objects#common_event_object

Reply:
@mtgto thank you! 

Re: helper - as far as I know, the above does not compile because one can only access the properties common to all members of the union, so the only way to use such a union is via narrowing with `"stringInputs" in bar`. Such a helper could look [like this](https://tsplay.dev/mbkD4W) avoiding the problem. But, anyway, that's just a passing thought, I too think the onus is more on the definition users.

Re: one of - you probably will not find anything concrete. I am interpreting this line in docs as proof of the "union field" behavior (+coming from experience - the real-world event objects confirm the supposition):

> The structure of the map value object is dependent on the widget type

I think it is a bit of a documentation shortcoming that it is not very clear (and I even forgot about that initially) :(

Reply:
@Oaphi Thanks for code example 👍  I understand how to write `OneOf` helper.

> I think it is a bit of a documentation shortcoming that it is not very clear

Me too. I also same opinion that it might appear only one of `xxxInputs` is set in actually even if doc does not declare it 😇

Reply:
@mtgto I only linked to the playground to illustrate what implementation I had in mind that avoided the "no fields in common" trap 😇 Ok, so it seems the only thing left to do is to add the folder.